### PR TITLE
[Scheduling Sample] Add bot to handle booking an appointment

### PR DESCRIPTION
**Pull request by pamella**
## Description

- Partial work of #4973 
- Update the scheduling flow that starts from My Schedule and Patient detail pages.

  | Starting Point | Flow |
  | -------------------- | ------- |
  | My Schedule | Click on an available slot -> Open appointment creation modal (slot field hidden) |
  | Patient detail | Click on create appointment action -> Redirect to a modified Scheduling page -> Select a slot -> Open appointment creation modal (slot and patient fields hidden)  |
  | My Appointments | Redirect to the My Schedule page (since it would be trickier to add the slots field to the create appointment modal questionnaire) |

- Add bot to handle booking an appointment
   - Move the logic from `CreateAppointment.handleQuestionnaireSubmit` to the `booking-appointment` bot.
   - Add tests for the bot code.

> [!NOTE]
> Other bots will be implemented in future PRs since this PR was already getting bigger due to the refactoring of the scheduling behavior flow.

## Steps to test

- [ ] Build the bots: `npm run build:bots`
- [ ] Upload example bots
- [ ] Create slots
  - [ ] Blocked
  - [ ] Available
- [ ] Create appointments
  - [ ] Starting from My Schedule page
  - [ ] Starting from Patient detail page
  - [ ] Starting from My Appointments page 

[Screencast from 15-08-2024 19:15:30.webm](https://github.com/user-attachments/assets/565d5928-0b55-405e-90f0-682efda0db59)

### CODE REVIEW

File: examples/medplum-scheduling-demo/src/App.tsx [hunk 1]
```diff
@@ -66,16 +68,20 @@ export function App(): JSX.Element | null {
...
             {
               icon: <IconCalendar />,
-              label: 'Schedule',
+              label: 'My Schedule',
               href: '/Schedule',
             },
             {
               icon: <IconClipboard />,
-              label: 'Appointments',
+              label: 'My Appointments',
```
Comment (pamella) [inline examples/medplum-scheduling-demo/src/App.tsx:76]: Rename for clarity.

File: examples/medplum-scheduling-demo/src/components/CreateAppointment.tsx [hunk 2]
```diff
@@ -92,30 +108,20 @@ const appointmentQuestionnaire: Questionnaire = {
...
-      type: 'dateTime',
-      text: 'End date',
-      required: true,
-    },
     {
       linkId: 'service-type',
       type: 'choice',
       text: 'What is the appointment service type?',
-      answerValueSet: 'http://hl7.org/fhir/ValueSet/service-type',
+      answerValueSet: 'http://terminology.hl7.org/ValueSet/v2-0276',
```
Comment (pamella) [inline examples/medplum-scheduling-demo/src/components/CreateAppointment.tsx:124]: Drive-by fix.

File: examples/medplum-scheduling-demo/src/components/CreateAppointment.tsx [hunk 3]
```diff
@@ -50,11 +61,16 @@ export function CreateAppointment(props: CreateAppointmentProps): JSX.Element {
             status: 'accepted',
           },
         ],
-      });
+      };
+      // Call bot to create the appointment
+      appointment = await medplum.executeBot({ system: 'http://example.com', value: 'book-appointment' }, appointment);
```
Comment (pamella) [inline examples/medplum-scheduling-demo/src/components/CreateAppointment.tsx:66]: The appointment is now being created via bot.

File: examples/medplum-scheduling-demo/src/pages/PatientSchedulePage.tsx
Comment (pamella) [inline examples/medplum-scheduling-demo/src/pages/PatientSchedulePage.tsx:1]: This new component manages the scheduling flow when initiated from the Patient detail page. It's similar to the `SchedulePage.tsx` (My Schedule page) but has been kept separate to avoid cluttering the existing code with additional logic.

File: examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx [hunk 4]
```diff
@@ -44,7 +44,17 @@ export function AppointmentActions(props: AppointmentActionsProps): JSX.Element
       await medplum.updateResource({
         ...appointment,
         status: newStatus,
+        slot: undefined,
       });
+      // If the appointment is cancelled, update the slot status to free
+      if (newStatus === 'cancelled' && appointment.slot?.[0].reference) {
+        const slotId = appointment.slot[0].reference.split('Slot/')[1];
```
Comment (rahul1) [inline examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx:51]: we have a helper function for this: `resolveId(reference)`

File: examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx [hunk 5]
```diff
@@ -44,7 +44,17 @@ export function AppointmentActions(props: AppointmentActionsProps): JSX.Element
...
+        const slot = await medplum.readResource('Slot', slotId);
+        await medplum.updateResource({
+          ...slot,
+          status: 'free',
+        });
+      }
```
Comment (rahul1) [inline examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx:57]: I think this cancellation logic should live in a bot rather than on the client. The reason is that, in the future, cancellation logic can get more complicated (e.g. also update a room to be free in a physical location. Send a notification, etc.)

File: examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx [hunk 6]
```diff
@@ -44,7 +44,17 @@ export function AppointmentActions(props: AppointmentActionsProps): JSX.Element
       await medplum.updateResource({
         ...appointment,
         status: newStatus,
```
Comment (rahul1) [inline examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx:46]: Use `PATCH` for these small, targetted changes

File: examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx [hunk 7]
```diff
@@ -44,7 +44,17 @@ export function AppointmentActions(props: AppointmentActionsProps): JSX.Element
...
+        slot: undefined,
       });
+      // If the appointment is cancelled, update the slot status to free
+      if (newStatus === 'cancelled' && appointment.slot?.[0].reference) {
+        const slotId = appointment.slot[0].reference.split('Slot/')[1];
+        const slot = await medplum.readResource('Slot', slotId);
+        await medplum.updateResource({
+          ...slot,
+          status: 'free',
```
Comment (rahul1) [inline examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx:55]: Use `PATCH` for these small, targeted changes

File: examples/medplum-scheduling-demo/src/pages/SchedulePage.tsx [hunk 8]
```diff
@@ -9,13 +9,40 @@ import 'react-big-calendar/lib/css/react-big-calendar.css';
...
+  readonly opened: boolean;
+  readonly handlers: {
+    readonly open: () => void;
+    readonly close: () => void;
+    readonly toggle: () => void;
+  };
+}
+
+// This helper component manage the modal to be shown when an event is clicked
+function OnClickEventModal(props: OnClickEventModalProps): JSX.Element {
```
Comment (rahul1) [inline examples/medplum-scheduling-demo/src/pages/SchedulePage.tsx:26]: Needs a clearer name

**Review by rahul1 — APPROVED** [review]

File: examples/medplum-scheduling-demo/src/pages/SchedulePage.tsx (hunk 8 above)
Comment (pamella) [reply examples/medplum-scheduling-demo/src/pages/SchedulePage.tsx:26]: Yes, I agree, I didn't like that name either. I'll refactor it.

File: examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx (hunk 4 above)
Comment (pamella) [reply examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx]: Nice, thanks!

File: examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx (hunk 6 above)
Comment (pamella) [reply examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx:47]: Alright, I'll do this refactoring in another PR when I move the code to a bot.

**Comment by pamella** [comment]
I'm merging the PR, but it's on my track to address the bot and patch notes in the upcoming PRs.
